src/fighter_state.rs

```diff
@@ -0,0 +1,747 @@
+}
+
+/// Component indicating the player is moving
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Moving {
+    pub velocity: Vec2,
+}
+impl Moving {
+    pub const PRIORITY: i32 = 10;
+    pub const ANIMATION: &'static str = "running";
+}
+
+/// The player is throwing an item
+#[derive(Component, Reflect, Default, Debug)]
+pub struct Throwing;
+impl Throwing {
+    pub const PRIORITY: i32 = 15;
+}
+
+/// The player is grabbing an item ( or trying to)
+#[derive(Component, Reflect, Default, Debug)]
+pub struct Grabbing;
+impl Grabbing {
+    pub const PRIORITY: i32 = Throwing::PRIORITY;
+}
+
+/// Component indicating the player is flopping
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Attacking {
+    /// The initial y-height of the figther when starting the attack
+    pub start_y: f32,
+    pub has_started: bool,
+    pub is_finished: bool,
+}
+impl Attacking {
+    pub const PRIORITY: i32 = 30;
+    pub const ANIMATION: &'static str = "attacking";
+}
+
+/// Component indicating the player is getting knocked back
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct KnockedBack {
+    pub velocity: Vec2,
+    pub timer: Timer,
+}
+impl KnockedBack {
+    pub const PRIORITY: i32 = 20;
+    pub const ANIMATION_LEFT: &'static str = "knocked_left";
+    pub const ANIMATION_RIGHT: &'static str = "knocked_right";
+}
+
+/// Component indicating the player is dying
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Dying;
+impl Dying {
+    pub const PRIORITY: i32 = 1000;
+    pub const ANIMATION: &'static str = "dying";
+}
+
+//
+// Fighter input collector systems
+//
+
+/// Emits state transitions based on fighter actions
+fn collect_player_actions(
+    mut players: Query<
+        (
+            &ActionState<PlayerAction>,
+            &mut StateTransitionIntents,
+            &Inventory,
+            &Stats,
+        ),
+        With<Player>,
+    >,
+) {
+    for (action_state, mut transition_intents, inventory, stats) in &mut players {
+        // Trigger attacks
+        if action_state.just_pressed(PlayerAction::Attack) {
+            transition_intents.push_back(StateTransition::new(
+                Attacking::default(),
+                Attacking::PRIORITY,
+                false,
+            ));
+        }
+
+        // Trigger grab/throw
+        if action_state.just_pressed(PlayerAction::Throw) {
+            if inventory.is_some() {
+                transition_intents.push_back(StateTransition::new(
+                    Throwing,
+                    Throwing::PRIORITY,
+                    true,
+                ));
+            } else {
+                transition_intents.push_back(StateTransition::new(
+                    Grabbing,
+                    Grabbing::PRIORITY,
+                    true,
+                ));
+            }
+        }
```

Comment:
intent triggering on input seems quite nice, liking how this looks :+1:

Reply:
So far it's working very nicely and was refreshingly easy to re-build the fighter state machine with the new model, piece by piece.

It felt easy to just come up with the next thing I wanted the fighter to be able to do, and it didn't require modifying a big centralized system of state management. I just had to add a new state, decide when to trigger it, and how to handle it.

---

I'm still a little unsure about what the best practice will be as far as where to make decisions about which state to transition to. For instance, right here we have the `collect_player_actions` system making a decision, based on the fighter's inventory, what state to transition to.

This seems good in this case, but it get's a little more confusing when consider things like mid-air attacks where the same attack button might trigger a different state. Should `collect_player_actions` be responsible for making that decision? Or does it just depend on a case-by-case basis?

We'll just have to think about options as we go I think and feel out how well it's working and what might be the best practice.

Reply:
yeahp im wondering that also but it doesnt need to be addressed here.
I was thinking about it in relation to combo/chain attacks and input buffering where you want to chain attacks/play animations based on sequence of buttons within certain frame windows.